HunFlair: Fixed links for biomedical datasets

WangXII opened the pull request:
This PR fixes links to datasets that are no longer available at their original location. Instead the links now target working data repositories and adjusts the data parsing scripts whenever necessary. Complements the changes in https://github.com/flairNLP/flair/pull/3030

alanakbik commented:
@WangXII could you give some more info on the parsing script changes? Also which corpora are affected?

WangXII commented:
@alanakbik BioInfer and IEPA are the only two corpora where I've adjusted the parser according to the xml format in the new links. For the rest, I've just replaced the old links and made sure building the conll files works.

The new xml format for BioInfer has simplified the parsing script as there is now only one xml tag per sentence and not one xml tag per token. The same format is now also used by IEPA which replaces the bioc parsing function from the old link.

Here is the difference in code:

<details>
  <summary>New version</summary>

``` py
@classmethod
    def parse_dataset(cls, original_file: Path):
        documents: Dict[str, str] = {}
        entities_per_document: Dict[str, List[Entity]] = {}

        tree = etree.parse(str(original_file))
        sentence_elems = tree.xpath("//sentence")
        for s_id, sentence in enumerate(sentence_elems):
            sentence_id = str(s_id)
            documents[sentence_id] = sentence.attrib["text"]
            entities_per_document[sentence_id] = []

            for entity in sentence.xpath(".//entity"):
                char_offsets = re.split("-|,", entity.attrib["charOffset"])
                start_token = int(char_offsets[0])
                end_token = int(char_offsets[-1])
                entities_per_document[sentence_id].append(
                    Entity(
                        char_span=(start_token, end_token),
                        entity_type=entity.attrib["type"],
                    )
                )

        return InternalBioNerDataset(documents=documents, entities_per_document=entities_per_document)
```
</details>

<details>
  <summary>Old version</summary>

``` py
@classmethod
    def parse_dataset(cls, original_file: Path):
        documents: Dict[str, str] = {}
        entities_per_document: Dict[str, List[Entity]] = {}

        tree = etree.parse(str(original_file))
        sentence_elems = tree.xpath("//sentence")
        for s_id, sentence in enumerate(sentence_elems):
            sentence_id = str(s_id)
            token_id_to_span = {}
            sentence_text = ""
            entities_per_document[sentence_id] = []

            for token in sentence.xpath(".//token"):
                token_text = "".join(token.xpath(".//subtoken/@text"))
                token_id = ".".join(token.attrib["id"].split(".")[1:])

                if not sentence_text:
                    token_id_to_span[token_id] = (0, len(token_text))
                    sentence_text = token_text
                else:
                    token_id_to_span[token_id] = (
                        len(sentence_text) + 1,
                        len(token_text) + len(sentence_text) + 1,
                    )
                    sentence_text += " " + token_text
            documents[sentence_id] = sentence_text

            entities = [
                e for e in sentence.xpath(".//entity") if not e.attrib["type"].isupper()
            ]  # all caps entity type apparently marks event trigger

            for entity in entities:
                token_nums = []
                entity_character_starts = []
                entity_character_ends = []

                for subtoken in entity.xpath(".//nestedsubtoken"):
                    token_id_parts = subtoken.attrib["id"].split(".")
                    token_id = ".".join(token_id_parts[1:3])

                    token_nums.append(int(token_id_parts[2]))
                    entity_character_starts.append(token_id_to_span[token_id][0])
                    entity_character_ends.append(token_id_to_span[token_id][1])

                if token_nums and entity_character_starts and entity_character_ends:
                    entity_tokens = list(zip(token_nums, entity_character_starts, entity_character_ends))

                    start_token = entity_tokens[0]
                    last_entity_token = entity_tokens[0]
                    for entity_token in entity_tokens[1:]:
                        if not (entity_token[0] - 1) == last_entity_token[0]:
                            entities_per_document[sentence_id].append(
                                Entity(
                                    char_span=(start_token[1], last_entity_token[2]),
                                    entity_type=entity.attrib["type"],
                                )
                            )
                            start_token = entity_token

                        last_entity_token = entity_token

                    if start_token:
                        entities_per_document[sentence_id].append(
                            Entity(
                                char_span=(start_token[1], last_entity_token[2]),
                                entity_type=entity.attrib["type"],
                            )
                        )

        return InternalBioNerDataset(documents=documents, entities_per_document=entities_per_document)
```
</details>

alanakbik commented on flair/datasets/biomedical.py:
```diff
@@ -188,7 +187,9 @@ def filter_nested_entities(dataset: InternalBioNerDataset) -> None:
     if num_entities_before != num_entities_after:
         removed = num_entities_before - num_entities_after
-        warn(f"Corpus modified by filtering nested entities. Removed {removed} entities.")
+        warn(
+            f"Corpus modified by filtering nested entities. Removed {removed} entities. Keep {num_entities_after} entities."
+        )
```
This warning prints oddly for me on first download. It prints 

```console
UserWarning: Corpus modified by filtering nested entities. Removed 63 entities. Keep 3548 entities.
  f"Corpus modified by filtering nested entities. Removed {removed} entities. Keep {num_entities_after} entities."
```

Not a big deal, but a bit strange that the f-string is printed as well. Can you check if this is happening on your end as well?

WangXII replied:
Yeah, happening on my end as well. Seems to depend the format of the warning messages, see https://stackoverflow.com/questions/2187269/print-only-the-message-on-warnings

I would just use logger.warning(...) instead of warnings.warn(...) as the logging tutorial recommends. The warning is an informative message and not a cause to modify the application, see https://stackoverflow.com/questions/9595009/warnings-warn-vs-logging-warning/14762106#14762106

How is this handled elsewhere in Flair?

alanakbik replied:
Yes, we use the logger for warnings normally (log.warning(..)). See for instance https://github.com/flairNLP/flair/blob/117e14e5d821a1a4a655ccbbe495fc380ecebb7a/flair/datasets/sequence_labeling.py#L827

WangXII replied:
Alright, I've changed the corresponding places in the code to logger.warning(...) instead.

alanakbik reviewed (approved):
Looks good, just one minor point regarding printouts.

alanakbik commented:
@WangXII thanks for fixing this! I believe the final black warning is fixed in master branch, so merging this!
